apis/compute/v1beta1/gkecluster_types.go

```diff
@@ -0,0 +1,1346 @@
+	//   "SET_BY_OPERATOR" - Cluster state was manually changed by an SRE
+	// due to a system logic error.
+	// More codes TBA
+	Code string `json:"code,omitempty"`
+
+	// Message: Human-friendly representation of the condition
+	Message string `json:"message,omitempty"`
+}
+
+// MaxPodsConstraint defines constraints applied to pods.
+type MaxPodsConstraint struct {
+	// MaxPodsPerNode: Constraint enforced on the max num of pods per node.
+	MaxPodsPerNode int64 `json:"maxPodsPerNode"`
+}
+
+// IPAllocationPolicy is configuration for controlling how IPs are
+// allocated in the cluster.
+type IPAllocationPolicy struct {
+	// AllowRouteOverlap: If true, allow allocation of cluster CIDR ranges
+	// that overlap with certain
+	// kinds of network routes. By default we do not allow cluster CIDR
+	// ranges to
+	// intersect with any user declared routes. With allow_route_overlap ==
+	// true,
+	// we allow overlapping with CIDR ranges that are larger than the
+	// cluster CIDR
+	// range.
+	//
+	// If this field is set to true, then cluster and services CIDRs must
+	// be
+	// fully-specified (e.g. `10.96.0.0/14`, but not `/14`), which means:
+	// 1) When `use_ip_aliases` is true, `cluster_ipv4_cidr_block` and
+	//    `services_ipv4_cidr_block` must be fully-specified.
+	// 2) When `use_ip_aliases` is false, `cluster.cluster_ipv4_cidr` muse
+	// be
+	//    fully-specified.
+	AllowRouteOverlap *bool `json:"allowRouteOverlap,omitempty"`
+
+	// ClusterIpv4CidrBlock: The IP address range for the cluster pod IPs. If
+	// this field is set, then `cluster.cluster_ipv4_cidr` must be left blank.
+	//
+	// This field is only applicable when `use_ip_aliases` is true.
+	//
+	// Set to blank to have a range chosen with the default size.
+	//
+	// Set to /netmask (e.g. `/14`) to have a range chosen with a specific
+	// netmask.
+	//
+	// Set to a
+	// [CIDR](http://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing)
+	//
+	// notation (e.g. `10.96.0.0/14`) from the RFC-1918 private networks (e.g.
+	// `10.0.0.0/8`, `172.16.0.0/12`, `192.168.0.0/16`) to pick a specific range
+	// to use.
+	// +optional
+	ClusterIpv4CidrBlock *string `json:"clusterIpv4CidrBlock,omitempty"`
+
+	// ClusterSecondaryRangeName: The name of the secondary range to be used
+	// for the cluster CIDR
+	// block.  The secondary range will be used for pod IP
+	// addresses. This must be an existing secondary range associated
+	// with the cluster subnetwork.
+	//
+	// This field is only applicable with use_ip_aliases is true
+	// and
+	// create_subnetwork is false.
+	// +optional
+	ClusterSecondaryRangeName *string `json:"clusterSecondaryRangeName,omitempty"`
+
+	// CreateSubnetwork: Whether a new subnetwork will be created
+	// automatically for the cluster.
+	//
+	// This field is only applicable when `use_ip_aliases` is true.
+	// TODO(hasheddan): should this be removed?
+	// +optional
+	CreateSubnetwork *bool `json:"createSubnetwork,omitempty"`
```

Comment:
We can allow subnetworks to be created for ip allocation, but would that be a violation of having subnetworks created outside of the `Subnetwork` resource? I would tend to think this is a special case that we could allow as part of the `GKECluster` configuration.

Reply:
Yeah I feel okay with allowing this, especially given it doesn't have any config or knobs for managing subnets short of "make one". I believe the subnet is tied to the lifecycle of the cluster in this case, too.